### PR TITLE
[RW-7732][risk=no] e2e test for User Profile Admin

### PR DESCRIPTION
**Pull request by aweng98**
Test pass on localhost.

Screen Shot 2022-02-16 at 1 15 10 PM

---
**PR checklist**

- [ ] This PR meets the Acceptance Criteria in the JIRA story
- [ ] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests
- [ ] I have run and tested this change locally
- [ ] I have run the E2E tests on ths change against my local UI and/or API server with `yarn test-local` or [yarn test-local-devup](https://github.com/all-of-us/workbench/blob/main/e2e/README.md#examples)
- [ ] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer
- [ ] If this includes an API change, I have updated the appropriate Swagger definitions and notified API consumers
- [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later

### CODE REVIEW

File: e2e/app/component/cell.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import Container from 'app/container';
+import { Page } from 'puppeteer';
+import { getPropValue } from 'utils/element-utils';
+
+export default class Cell extends Container {
```
Comment (aweng98) [inline e2e/app/component/cell.ts:5]: I plan to make a follow up PR to move `table.ts` and `cell.ts` from `component` dir to new `table` dir. At some point in future, new `column.ts` will be added.

File: e2e/app/component/table.ts [hunk 2]
```diff
@@ -90,11 +99,50 @@ export default class Table extends Container {
...
+    for (let i = 0; i < rows.length; i++) {
+      const tds = await rows[i].$x('./td');
+      for (let j = 0; j < tds.length; j++) {
+        const v = await getPropValue<string>(tds[j], 'innerText');
+        if (v === rowValue) {
+          rowIndex = i;
+          break;
+        }
+      }
+    }
```
Comment (aweng98) [inline e2e/app/component/table.ts:115]: I could use some help to rewrite this block of code to functional code.

File: e2e/app/element/switch.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
+import BaseElement from './base-element';
+import { Page } from 'puppeteer';
+import { ElementType, XPathOptions } from 'app/xpath-options';
+import Container from 'app/container';
+import { buildXPath } from 'app/xpath-builders';
+import { getPropValue } from 'utils/element-utils';
+
+export const defaultSwitchXpath = '//label[.//input[@type="checkbox" and @role="switch"]]';
+
+export default class Switch extends BaseElement {
```
Comment (aweng98) [inline e2e/app/element/switch.ts:10]: `Toggle` or `Switch`. Any one prefers one over another?

File: e2e/app/page/cohort-build-page.ts [hunk 4]
```diff
@@ -73,7 +73,7 @@ export default class CohortBuildPage extends AuthenticatedPage {
       default:
         break;
     }
-    await waitForText(this.page, 'Cohort Saved Successfully');
```
Comment (jmthibault79) [inline e2e/app/page/cohort-build-page.ts:76]: why this change?

File: e2e/app/page/create-account-page.ts [hunk 5]
```diff
@@ -200,7 +191,7 @@ export default class CreateAccountPage extends BasePage {
     }).asElementHandle();
 
     const roleSelect = SelectMenu.findByName(this.page, FieldSelector.DescribeRole.textOption);
-    await roleSelect.select(InstitutionRoleSelectValue.UndergraduteStudent);
+    await roleSelect.select(InstitutionRole.UndergraduteStudent);
```
Comment (jmthibault79) [inline e2e/app/page/create-account-page.ts:194]: spelling: undergraduate

File: e2e/utils/test-utils.ts [hunk 6]
```diff
@@ -334,3 +334,7 @@ export async function openTab<T extends AuthenticatedPage>(page: Page, tabName:
     await tab.waitFor(pageExpected);
   }
 }
+
+export function parseForNumbericalString(text: string): RegExpMatchArray | null {
```
Comment (jmthibault79) [inline e2e/utils/test-utils.ts:338]: spelling: numerical

File: e2e/tests/nightly/admin/user-profile-admin.spec.ts [hunk 7]
```diff
@@ -0,0 +1,469 @@
...
+  ];
+
+  const testUserEmail = 'admin_test';
+  let adminTab: Page;
+
+  beforeEach(async () => {
+    adminTab = await searchTestUser(page, testUserEmail);
+  });
+
+  test('Verify Researcher Information renders correctly', async () => {
```
Comment (jmthibault79) [inline e2e/tests/nightly/admin/user-profile-admin.spec.ts:66]: also Edit information

File: e2e/tests/nightly/admin/user-profile-admin.spec.ts [hunk 8]
```diff
@@ -0,0 +1,469 @@
...
+    expect(name).toEqual('Admin Testing');
+
+    // Verify User Name
+    const userName = await userProfileAdminPage.getUserName();
+    expect(userName).toEqual(`${testUserEmail}${config.EMAIL_DOMAIN_NAME}`);
+
+    // Verify Credits Used
+    const [usedAmount, creditsLimitAmount] = await userProfileAdminPage.getInitialCreditsUsed();
+    expect(usedAmount).toEqual(0);
+    expect(creditsLimitAmount).toEqual(credits); // Matches Initial Credit Limit selected value
```
Comment (jmthibault79) [inline e2e/tests/nightly/admin/user-profile-admin.spec.ts:126]: nice

File: e2e/tests/nightly/admin/user-profile-admin.spec.ts [hunk 9]
```diff
@@ -0,0 +1,469 @@
...
+
+  test('Verify Access Status table render correctly', async () => {
+    const userProfileAdminPage = new UserProfileAdminPage(adminTab);
+    await userProfileAdminPage.waitForLoad();
+
+    // Verify column headers displayed correctly
+    const accountAccessTable = userProfileAdminPage.getAccessStatusTable();
+    const columnNames = await accountAccessTable.getColumnNames();
+    expect(columnNames.sort()).toEqual(accessStatusTableColumns.sort());
+    expect(columnNames.length).toEqual(accessStatusTableColumns.length);
```
Comment (jmthibault79) [inline e2e/tests/nightly/admin/user-profile-admin.spec.ts:137]: don't need to check length because you already know the contents match

File: e2e/tests/nightly/admin/user-profile-admin.spec.ts [hunk 10]
```diff
@@ -0,0 +1,469 @@
...
+
+    const saveButton = userProfileAdminPage.getSaveButton();
+    const cancelButton = userProfileAdminPage.getCancelButton();
+    await verifyButtonEnabled(cancelButton, false);
+    await verifyButtonEnabled(saveButton, false);
+
+    // Turn off Account Access toggle
+    await accountAccessSwitch.turnOff();
+    label = await accountAccessSwitch.getLabel();
+    expect(label).toEqual('Account disabled');
```
Comment (jmthibault79) [inline e2e/tests/nightly/admin/user-profile-admin.spec.ts:166]: Cancel and Save should be enabled now, so let's check those

File: e2e/tests/nightly/admin/user-profile-admin.spec.ts [hunk 11]
```diff
@@ -0,0 +1,469 @@
...
+    const columnNames = await accountAccessTable.getColumnNames();
+    expect(columnNames.sort()).toEqual(accessStatusTableColumns.sort());
+    expect(columnNames.length).toEqual(accessStatusTableColumns.length);
+
+    // Verify Access module names displayed correctly
+    const moduleNames = await accountAccessTable.getRowValues(1);
+    expect(moduleNames.sort()).toEqual(accessModules.sort());
+    expect(moduleNames.length).toEqual(accessModules.length);
+
+    // check badges
```
Comment (jmthibault79) [inline e2e/tests/nightly/admin/user-profile-admin.spec.ts:144]: reminder to do this

File: e2e/tests/nightly/admin/user-profile-admin.spec.ts [hunk 12]
```diff
@@ -0,0 +1,469 @@
...
+  await new UserProfileInfo(newPage).waitForLoad();
+
+  // Go to url users-tmp page
+  const tmpUrl = replaceWithTmpUrl(newPage.url());
+  await newPage.goto(tmpUrl, { waitUntil: ['load', 'networkidle0'] });
+
+  return newPage;
+}
+
+async function verifyInstitutionAgreementPage(page: Page): Promise<void> {
```
Comment (jmthibault79) [inline e2e/tests/nightly/admin/user-profile-admin.spec.ts:453]: interesting ... I hadn't clicked this before

File: e2e/tests/nightly/admin/user-profile-admin.spec.ts [hunk 13]
```diff
@@ -0,0 +1,469 @@
...
+
+    // Find email error
+    const emailErrorMsg = await userProfileAdminPage.getEmailErrorMessage();
+    expect(emailErrorMsg).toContain('The institution has authorized access only to select members');
+
+    await verifyInstitutionAgreementPage(adminTab);
+    await adminTab.bringToFront();
+
+    await verifyButtonEnabled(cancelButton, true);
+    // SAVE buttons is disabled due to error in page
```
Comment (jmthibault79) [inline e2e/tests/nightly/admin/user-profile-admin.spec.ts:216]: good

File: e2e/tests/nightly/admin/user-profile-admin.spec.ts [hunk 14]
```diff
@@ -0,0 +1,469 @@
...
+
+    // Type in modified email
+    const invalidContactEmail = 'mod-' + oldContactEmail;
+    await contactEmail.type(invalidContactEmail);
+
+    // Input background color changed
+    backgroundColor = await getStyleValue<string>(adminTab, await contactEmail.asElementHandle(), 'background-color');
+    expect(backgroundColor).toEqual('rgb(248, 201, 84)');
+
+    // Find email error
```
Comment (jmthibault79) [inline e2e/tests/nightly/admin/user-profile-admin.spec.ts:208]: add something about how this is the erroor message we see when the Inst matches exact email addresses

File: e2e/tests/nightly/admin/user-profile-admin.spec.ts [hunk 15]
```diff
@@ -0,0 +1,469 @@
...
+    await verifiedInstitution.select('Broad Institute');
+
+    // Check error when Verified Institution and Contact Email domains do not match
+    await waitForText(adminTab, 'Your email does not match your institution');
+
+    await verifyButtonEnabled(cancelButton, true);
+    // SAVE buttons is disabled due to error in page
+    await verifyButtonEnabled(saveButton, false);
+
+    // After click CANCEL button, unsaved changes are automatically discarded
```
Comment (jmthibault79) [inline e2e/tests/nightly/admin/user-profile-admin.spec.ts:235]: cool, I like this test

File: e2e/tests/nightly/admin/user-profile-admin.spec.ts [hunk 16]
```diff
@@ -0,0 +1,469 @@
...
+  });
+
+  test('Can bypass Google 2-Step Verification module', async () => {
+    const userProfileAdminPage = new UserProfileAdminPage(adminTab);
+    await userProfileAdminPage.waitForLoad();
+
+    const accessStatusTable = await userProfileAdminPage.getAccessStatusTable();
+
+    // Last completed on is a dash
+    const lastCompletedOnCell = await accessStatusTable.getCellByValue(
```
Comment (jmthibault79) [inline e2e/tests/nightly/admin/user-profile-admin.spec.ts:274]: oh this function is cool

File: e2e/tests/nightly/admin/user-profile-admin.spec.ts [hunk 17]
```diff
@@ -0,0 +1,469 @@
...
+      TableColumns.EXPIRES_ON
+    );
+    const expiresOn = await expiresOnCell.getCellValue();
+    expect(expiresOn).toEqual('Never');
+
+    // Change Bypass for Google 2-step verification module
+    const statusCell = await accessStatusTable.getCellByValue(AccessModules.GOOGLE_VERIFICATION, TableColumns.STATUS);
+    const oldStatus = await statusCell.getCellValue();
+
+    const bypassSwitch = await userProfileAdminPage.getBypassSwitchForRow(AccessModules.GOOGLE_VERIFICATION);
```
Comment (jmthibault79) [inline e2e/tests/nightly/admin/user-profile-admin.spec.ts:293]: also cool

File: e2e/tests/nightly/admin/user-profile-admin.spec.ts [hunk 18]
```diff
@@ -0,0 +1,469 @@
...
+    }
+
+    if (newStatus === 'Incomplete') {
+      await bypassSwitch.turnOn();
+      await saveButton.click();
+      await userProfileAdminPage.waitForLoad();
+    }
+  });
+
+  test('Can bypass Controlled Tier training module', async () => {
```
Comment (jmthibault79) [inline e2e/tests/nightly/admin/user-profile-admin.spec.ts:329]: curious - why these 2 modules?

File: e2e/tests/nightly/admin/user-profile-admin.spec.ts [hunk 19]
```diff
@@ -0,0 +1,469 @@
...
+async function verifyButtonEnabled(button: Button, enabled: boolean): Promise<void> {
+  expect(await button.isCursorNotAllowed()).toBe(!enabled);
+}
+
+async function verifyToggleSwitchNewBackgroundColor(page: Page, cell: Cell): Promise<void> {
+  const divXpath = `${cell.getXpath()}/div`;
+
+  const background = await getStyleValue<string>(page, await page.waitForXPath(divXpath), 'background-color');
+
+  expect(background).toEqual('rgb(248, 201, 84)');
```
Comment (jmthibault79) [inline e2e/tests/nightly/admin/user-profile-admin.spec.ts:423]: this would make a good const, like CHANGED_COLOR or something

File: e2e/tests/nightly/admin/user-profile-admin.spec.ts [hunk 20]
```diff
@@ -0,0 +1,469 @@
...
+  test('Can bypass Controlled Tier training module', async () => {
+    const userProfileAdminPage = new UserProfileAdminPage(adminTab);
+    await userProfileAdminPage.waitForLoad();
+
+    const accessStatusTable = await userProfileAdminPage.getAccessStatusTable();
+    const expiresOnCell = await accessStatusTable.getCellByValue(AccessModules.CT_TRAINING, TableColumns.EXPIRES_ON);
+    const expiresOn = await expiresOnCell.getCellValue();
+    expect(expiresOn).toEqual('-');
+
+    // Change Bypass for Google 2-step verification module
```
Comment (jmthibault79) [inline e2e/tests/nightly/admin/user-profile-admin.spec.ts:338]: copypaste

File: e2e/tests/nightly/admin/user-profile-admin.spec.ts [hunk 21]
```diff
@@ -0,0 +1,469 @@
...
+    const statusCell = await accessStatusTable.getCellByValue(AccessModules.CT_TRAINING, TableColumns.STATUS);
+    const oldStatus = await statusCell.getCellValue();
+
+    const bypassSwitch = await userProfileAdminPage.getBypassSwitchForRow(AccessModules.CT_TRAINING);
+    if (oldStatus === 'Bypassed') {
+      expect(await userProfileAdminPage.getDataAccessTiers()).toEqual('Registered Tier');
```
Comment (jmthibault79) [inline e2e/tests/nightly/admin/user-profile-admin.spec.ts:344]: why not Controlled Tier here?  that would be a more interesting test

File: e2e/tests/nightly/admin/user-profile-admin.spec.ts [hunk 22]
```diff
@@ -0,0 +1,469 @@
...
+    const userProfileAdminPage = new UserProfileAdminPage(adminTab);
+    await userProfileAdminPage.waitForLoad();
+
+    // Modify Institution Role
+    const institutionRole = userProfileAdminPage.getInstitutionalRole();
+    const oldRole = await institutionRole.getSelectedValue();
+
+    let options = (await institutionRole.getAllOptionTexts()).filter((role) => role !== oldRole);
+    const randRole = fp.shuffle(options)[0];
+    await institutionRole.select(randRole);
```
Comment (jmthibault79) [inline e2e/tests/nightly/admin/user-profile-admin.spec.ts:385]: careful here - if it chooses Other, then you have to enter the Other Text also.  May want to remove that choice.

File: e2e/tests/nightly/admin/user-profile-admin.spec.ts [hunk 23]
```diff
@@ -0,0 +1,469 @@
...
+
+    const newRole = await institutionRole.getSelectedValue();
+    expect(newRole).toEqual(randRole);
+
+    const newCreditLimit = await initialCreditLimit.getSelectedValue();
+    expect(newCreditLimit).toEqual(randCreditLimit);
+  });
+});
+
+async function verifyButtonEnabled(button: Button, enabled: boolean): Promise<void> {
```
Comment (jmthibault79) [inline e2e/tests/nightly/admin/user-profile-admin.spec.ts:414]: This would be nice to have in the Button class, so other tests can use it

Comment (jmthibault79) [reply e2e/tests/nightly/admin/user-profile-admin.spec.ts:414]: or BaseElement?

**Review by jmthibault79 — COMMENTED** [review]
Pretty good test, thanks!

But I have a few comments

File: e2e/tests/nightly/admin/user-profile-admin.spec.ts (hunk 19 above)
Comment (aweng98) [reply e2e/tests/nightly/admin/user-profile-admin.spec.ts]: yes, `const CHANGED_BACKGROUND_COLOR`

File: e2e/tests/nightly/admin/user-profile-admin.spec.ts (hunk 18 above)
Comment (jmthibault79) [reply e2e/tests/nightly/admin/user-profile-admin.spec.ts:364]: Hm - not sure I agree.  But we can have a conversation about that for a different PR.  For this one, could you add something like "chose these 2 arbitrarily"

File: e2e/tests/nightly/admin/user-profile-admin.spec.ts (hunk 21 above)
Comment (jmthibault79) [reply e2e/tests/nightly/admin/user-profile-admin.spec.ts:379]: We need to have an institution on Test that remains in an expected state for testing.  How about something like "For e2e Testing - DO NOT CHANGE MANUALLY"

File: e2e/tests/nightly/admin/user-profile-admin.spec.ts (hunk 22 above)
Comment (jmthibault79) [reply e2e/tests/nightly/admin/user-profile-admin.spec.ts:422]: great!  Could you add a comment explaining why we did this?

File: e2e/app/component/table.ts (hunk 2 above)
Comment (jmthibault79) [reply e2e/app/component/table.ts:115]: OK - this would be a fun mob activity.  Let's merge the PR first, though

File: e2e/tests/nightly/admin/user-profile-admin.spec.ts [hunk 24]
```diff
@@ -0,0 +1,507 @@
...
+    // Click name to navigate to User Profile Admin page
+    await userAdminPage.clickUserName(1, nameColumnIndex);
+
+    const newTarget = await browser.waitForTarget((target) => target.opener() === page.target());
+    const newPage = await newTarget.page();
+
+    await new UserProfileInfo(newPage).waitForLoad();
+
+    // Go to url users-tmp page
+    const tmpUrl = replaceWithTmpUrl(newPage.url());
```
Comment (jmthibault79) [inline e2e/tests/nightly/admin/user-profile-admin.spec.ts:477]: I like how this is nicely localized, to make it easy to remove later

File: e2e/tests/nightly/admin/user-profile-admin.spec.ts [hunk 25]
```diff
@@ -0,0 +1,507 @@
...
+        case AccessModules.LOGIN_GOV:
+          // Match Never
+          expect(await cell.getText()).toEqual('Never');
+          break;
+        case AccessModules.UPDATE_PROFILE:
+        case AccessModules.PUBLICATION:
+          // Match date
+          expect(await cell.getText()).toEqual('May 12, 2031');
+          break;
+        default:
```
Comment (jmthibault79) [inline e2e/tests/nightly/admin/user-profile-admin.spec.ts:182]: we can check the text for the remaining modules, too.  It should be "-" right?

File: e2e/tests/nightly/admin/user-profile-admin.spec.ts [hunk 26]
```diff
@@ -0,0 +1,507 @@
...
+
+    // Change Initial Credits Limit to $400
+    const initialCreditLimit = userProfileAdminPage.getInitialCreditLimit();
+    const oldCreditLimit = await initialCreditLimit.getSelectedValue();
+    await initialCreditLimit.select('$400.00');
+
+    // Verify background color in Select has changed
+    const [styleElement] = await (await initialCreditLimit.asElement()).$x('./preceding-sibling::style');
+    const styleText = await getPropValue<string>(styleElement, 'innerText');
+    expect(styleText).toContain('background-color: #F8C954;');
```
Comment (jmthibault79) [inline e2e/tests/nightly/admin/user-profile-admin.spec.ts:226]: also a const

File: e2e/tests/nightly/admin/user-profile-admin.spec.ts [hunk 27]
```diff
@@ -0,0 +1,507 @@
...
+    // Change Contact Email
+    const contactEmail = userProfileAdminPage.getContactEmail();
+    const oldContactEmail = await contactEmail.getValue();
+    // Input background color is white before change
+    let backgroundColor = await getStyleValue<string>(
+      adminTab,
+      await contactEmail.asElementHandle(),
+      'background-color'
+    );
+    expect(backgroundColor).toEqual('rgb(255, 255, 255)');
```
Comment (jmthibault79) [inline e2e/tests/nightly/admin/user-profile-admin.spec.ts:240]: const

File: e2e/tests/nightly/admin/user-profile-admin.spec.ts [hunk 28]
```diff
@@ -0,0 +1,507 @@
...
+  });
+
+  test('Can go to User Audit page', async () => {
+    const userProfileAdminPage = new UserProfileAdminPage(adminTab);
+    await userProfileAdminPage.waitForLoad();
+
+    // Check AUDIT link is working
+    await userProfileAdminPage.getAuditLink().click();
+    const newTarget = await browser.waitForTarget((target) => target.opener() === page.target());
+    const newPage2 = await newTarget.page();
```
Comment (jmthibault79) [inline e2e/tests/nightly/admin/user-profile-admin.spec.ts:299]: why 2?

**Review by jmthibault79 — APPROVED** [review]
Looks good!  No real blockers here.

Re: Toggle vs Switch - our component `Toggle` uses the React `Switch` internally, so Switch is a good name for this.
